pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java

```diff
@@ -94,51 +94,157 @@ public TransformResultMetadata getResultMetadata() {
+      }
+      int numDocs = projectionBlock.getNumDocs();
+      if (resultStoredType == DataType.INT) {
+        int[] intValues = _transformFunction.transformToIntValuesSV(projectionBlock);
+        ArrayCopyUtils.copy(intValues, _longValuesSV, numDocs);
+      } else if (resultStoredType == DataType.FLOAT) {
+        float[] floatValues = _transformFunction.transformToFloatValuesSV(projectionBlock);
+        ArrayCopyUtils.copy(floatValues, _longValuesSV, numDocs);
+      } else {
+        double[] doubleValues = _transformFunction.transformToDoubleValuesSV(projectionBlock);
+        ArrayCopyUtils.copy(doubleValues, _longValuesSV, numDocs);
+      }
+      return _longValuesSV;
+    }
   }
 
   @Override
   public float[] transformToFloatValuesSV(ProjectionBlock projectionBlock) {
-    return _transformFunction.transformToFloatValuesSV(projectionBlock);
+    // When casting to INT/LONG, need to first read as the result type then convert to float values
+    DataType resultStoredType = _resultMetadata.getDataType().getStoredType();
+    if (resultStoredType != DataType.INT && resultStoredType != DataType.LONG) {
+      return _transformFunction.transformToFloatValuesSV(projectionBlock);
+    } else {
+      if (_floatValuesSV == null) {
+        _floatValuesSV = new float[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+      }
+      int numDocs = projectionBlock.getNumDocs();
+      if (resultStoredType == DataType.INT) {
+        int[] intValues = _transformFunction.transformToIntValuesSV(projectionBlock);
+        ArrayCopyUtils.copy(intValues, _floatValuesSV, numDocs);
+      } else {
+        long[] longValues = _transformFunction.transformToLongValuesSV(projectionBlock);
+        ArrayCopyUtils.copy(longValues, _floatValuesSV, numDocs);
+      }
+      return _floatValuesSV;
+    }
   }
 
   @Override
   public double[] transformToDoubleValuesSV(ProjectionBlock projectionBlock) {
-    return _transformFunction.transformToDoubleValuesSV(projectionBlock);
+    // When casting to INT/LONG/FLOAT, need to first read as the result type then convert to double values
+    DataType resultStoredType = _resultMetadata.getDataType().getStoredType();
+    if (resultStoredType != DataType.INT && resultStoredType != DataType.LONG && resultStoredType != DataType.FLOAT) {
+      return _transformFunction.transformToDoubleValuesSV(projectionBlock);
+    } else {
+      if (_doubleValuesSV == null) {
+        _doubleValuesSV = new double[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+      }
+      int numDocs = projectionBlock.getNumDocs();
+      if (resultStoredType == DataType.INT) {
+        int[] intValues = _transformFunction.transformToIntValuesSV(projectionBlock);
+        ArrayCopyUtils.copy(intValues, _doubleValuesSV, numDocs);
+      } else if (resultStoredType == DataType.LONG) {
+        long[] longValues = _transformFunction.transformToLongValuesSV(projectionBlock);
+        ArrayCopyUtils.copy(longValues, _doubleValuesSV, numDocs);
+      } else {
```

Comment:
do we need to handle float type?

Reply:
Yes, and that is in the last `else` block